lib/api/exception.dart

```diff
@@ -11,13 +13,26 @@ sealed class ApiRequestException implements Exception {
   /// and its route name would be "getMessages".
   final String routeName;
 
-  /// A user-facing description of the error.
+  /// Server supplied user-facing description of the error.
+  ///
+  /// For [ZulipApiException] this may be supplied by the server as the `message`
+  /// property in the JSON response.
+  // TODO(i18n): no method yet to translate a server provided string.
+  final String? message;
+
+  /// App supplied user-facing description of the error.
   ///
-  /// For [ZulipApiException] this is supplied by the server as the `message`
-  /// property in the JSON response, and is translated into the user's language.
-  final String message;
+  /// This description will be translated according to the
+  /// current set locale.
+  final String Function(ZulipLocalizations)? translatableMessage;
+
+  ApiRequestException({required this.routeName, required this.message, required this.translatableMessage});
 
-  ApiRequestException({required this.routeName, required this.message});
+  String toTranslatedString(ZulipLocalizations zulipLocalizations) {
+    return (translatableMessage != null)
+      ? translatableMessage!(zulipLocalizations)
+      : super.toString();
+  }
```

Comment:
I kind of think that this should become the only public way to read the error message, so that the two fields that back it would become private. It seems like in this design, it'd be a bug if we ever take the plain `message` and show that to the user, right? And that'd be a very easy bug to have.

In fact, asking the IDE to show references to the field, it seems like in this version of the PR we already introduce that bug — there's a reference to `e.message` in `lib/widgets/login.dart`:
```dart
        final message = (e is ZulipApiException)
          ? zulipLocalizations.loginServerErrorMessage(e.toTranslatedString(zulipLocalizations))
          : e.message;
        showErrorDialog(context: context,
          title: zulipLocalizations.loginServerLoginFailed,
          message: message);
```

Then the name of this method can also become something simpler, like `message`.